### PR TITLE
{Role} Make test_role_assignment_audits live run more reliable

### PR DESCRIPTION
**Pull request by jiasli**
**Description**  
According to the [live run result](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=537774&view=ms.vss-test-web.build-test-results-tab), the test `test_role_assignment_audits` is still unstable during live run:

```py
>               self.assertTrue([x for x in result if (resource_group in x['scope'] and
                                                       x['principalName'] == self.kwargs['upn'])])
E                                                      AssertionError: [] is not true

azure-cli/src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py:575: AssertionError
```

This is because the activity log from `az role assignment list-changelogs` doesn't appear immediately after `az role assignment create` finishes. The `sleep` statement is very experiential and may not be sufficient.

This PRs makes the test more reliable by **adopting a retry logic to continuously check whether the assertions are met**.

**Testing Guide**  
```
azdev test test_role_assignment_audits
```

### CODE REVIEW

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -330,73 +330,75 @@ def list_role_assignment_change_logs(cmd, start_time=None, end_time=None):  # py
...
-                    entry['scope'] = e.authorization.scope
-                elif s.http_request.method == 'DELETE':
-                    payload = e.properties and e.properties.get('responseBody')
-                    entry['action'] = 'Revoked'
-            if payload:
-                try:
-                    payload = json.loads(payload)
-                except ValueError:
-                    pass
+    if start_events:
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:333]: Check if `start_events` has some value to reduce unnecessary calls to 
- Role Definitions API
- Graph Object API

to 
- Improve the performance of the command
- Reduce the size of recording YAML

File: src/azure-cli/azure/cli/command_modules/role/tests/util.py [hunk 2]
```diff
@@ -9,21 +9,21 @@
 ROLE_COMMAND_SLEEP_DURATION = 10
 
 
-def cmd_with_retry(self, *args, sleep_duration=ROLE_COMMAND_SLEEP_DURATION, max_retry=ROLE_COMMAND_MAX_RETRY, **kwargs):
-    """Retry self.cmd and checks until success."""
+def retry(func, sleep_duration=ROLE_COMMAND_SLEEP_DURATION, max_retry=ROLE_COMMAND_MAX_RETRY):
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/util.py:12]: This function is refactored to support retrying a code block (enclosed within a function).

File: src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py [hunk 3]
```diff
@@ -554,26 +554,28 @@ def test_role_assignment_audits(self, resource_group):
 
                 if self.is_live or self.in_recording:
                     now = datetime.datetime.utcnow()
-                    start_time = '{}-{}-{}T{}:{}:{}Z'.format(now.year, now.month, now.day - 1, now.hour,
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py:557]: `now.day - 1` will fail on the first day of each month.

File: src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py [hunk 4]
```diff
@@ -554,26 +554,28 @@ def test_role_assignment_audits(self, resource_group):
...
-                                                             now.minute, now.second)
-                    time.sleep(120)
-                    result = self.cmd('role assignment list-changelogs --start-time {}'.format(start_time)).get_output_in_json()
+                    start = now - datetime.timedelta(minutes=1)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py:557]: Narrow down the time windows to reduce the size of the recording.

File: src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py [hunk 5]
```diff
@@ -554,26 +554,28 @@ def test_role_assignment_audits(self, resource_group):
...
+                    end = now + datetime.timedelta(minutes=1)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/latest/test_role.py:558]: Add one more minute to leave some margin.

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 6]
```diff
@@ -330,73 +330,75 @@ def list_role_assignment_change_logs(cmd, start_time=None, end_time=None):  # py
...
+                            if index != -1:
+                                entry['scope'] = entry['scope'][:index]
+                            parts = list(filter(None, entry['scope'].split('/')))
+                            entry['scopeName'] = parts[-1]
+                            if len(parts) < 3:
+                                entry['scopeType'] = 'Subscription'
+                            elif len(parts) < 5:
+                                entry['scopeType'] = 'Resource group'
+                            else:
+                                entry['scopeType'] = 'Resource'
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:383]: Too many indents. Too long. Hard to read. Maybe you need to create some functions.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:383]: True, but I intent not to change the code unless absolutely necessary.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:383]: True, but I intend not to change the code unless absolutely necessary.

**Comment by yonzhan** [comment]
Role

**Review by qwordy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/role/custom.py (hunk 1 above)
Comment (arrownj) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:333]: It seems that this change have nothing to do with the PR title, right ?

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py:333]: It has no impact on the functionality either. If `start_events` is `[]`, the result of `list_role_definitions` will be discarded.

**Review by arrownj — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]

**Pull request merged**